Add support for ...WithAttributes functions

jls5177 opened the pull request:
Implements #41 by extending the changes from the with-attributes branch. Most of the change is just boiler plate code to the various key generation functions. The bulk of the changes is to the new helper functions and types to make working with Attributes easier. 

**Summary of new changes:**
- Converted the Attribute type into an alias for the pkcs11.Attribute type to allow for direct conversion between the 2 types. Did the same for the AttributeType type.
- Added NewAttributewith... functions to simplify the common attribute generation functions
- Created an AttributeSet type to provide some common APIs (Add, Append, Merge, Copy, AddDefaults) when working with Attributes. These APIs return a pointer to the struct to allow for chaining the methods.
- Removed the code that returned the modified Attribute slices since the new AttributeSet is mutable and will contain the added Attributes after the call.

Commit:
WIP commit, partial implementation

This change was cherry-picked from the with-attributes branch from
the upstream repo, ThalesIgnite/crypto11.

commit 8f08aab76caa20308917a3e40efd8a7ac8f225e2

jls5177 committed:
feat: finish WithAttributes implementation

Added a WithAttributes to DSA, ECDSA, and Symmetric algorithms. Added
a few helper functions to minimize the boiler plate code needed to
add ID and Label Attributes since this is required for all crypto
algorithms.

jls5177 committed:
chore: cleanup attributes by adding new types

dmjones commented on crypto11.go:
```diff
@@ -310,7 +310,9 @@ func Configure(config *Config) (*Context, error) {
-	if err != nil {
+
+	// the login will fail when using a module slot so only error out when the token requires a login
+	if err != nil && (instance.token.Flags&pkcs11.CKF_LOGIN_REQUIRED) == pkcs11.CKF_LOGIN_REQUIRED {
```
This is an unrelated change, so I'd prefer we separated this out.

I'd also like to discuss the best way to approach this. I haven't tested whether CKF_LOGIN_REQUIRED is accurately populated by popular PKCS#11 implementations. An alternative might be to have a config setting controlling whether login is required or not.

jls5177 replied:
I agree, I created #46 to discuss how to proceed with this issue.

dmjones commented on attributes.go:
```diff
@@ -0,0 +1,256 @@
+	return a
+}
+
+func (a *AttributeSet) Append(attribute *Attribute) *AttributeSet {
```
Given the operation of the function, would `Set` be a more appropriate name? Or something else that makes it clear an existing attribute of the same type is overridden by this.

jls5177 replied:
I agree with the name `Set()`. This will be changed in the next patchset.

dmjones commented on attributes.go:
```diff
@@ -0,0 +1,256 @@
+	return a
+}
+
+func (a *AttributeSet) AddDefaultAttributes(additional []*Attribute) *AttributeSet {
```
The name of this function doesn't seem to match the logic. Something like `AddIfNotPresent` (or a more succinct version) might fit better.

jls5177 replied:
Done

dmjones commented on attributes.go:
```diff
@@ -0,0 +1,256 @@
+func (a *AttributeSet) Copy() *AttributeSet {
+	b := NewAttributeSet()
+	for _, v := range a.Attributes {
+		b.Attributes[v.Type] = v
```
Should this be a deep copy? If not, the docs for this function will need to make it clear its shallow.

jls5177 replied:
Yeah there is no real reason to return a shallow copy. I will update the code.

dmjones commented on attributes.go:
```diff
@@ -0,0 +1,256 @@
+
+// NewAttributewithLabel is a helper function that populates a new slice of Attributes with the provided ID and Label.
+// This function returns an error if either the ID or the Label is an empty slice.
+func NewAttributewithLabel(id, label []byte) (a *AttributeSet, err error) {
```
Should this be called `NewAttributeSetWithIDAndLabel`?  (note the `Set` and `WithID`)?

jls5177 replied:
Done

dmjones commented on attributes.go:
```diff
@@ -0,0 +1,256 @@
+
+// NewAttributewithId is a helper function that populates a new slice of Attributes with the provided ID.
+// This function returns an error if the ID is an empty slice.
+func NewAttributewithId(id []byte) (*AttributeSet, error) {
```
Should this be `NewAttributeSetWithID` (note: the `Set` as well as capitalisation)?

jls5177 replied:
Yes, I agree with the naming convention.

dmjones commented on attributes.go:
```diff
@@ -0,0 +1,256 @@
+}
+
+func (a *AttributeSet) Add(attributeType AttributeType, value interface{}) *AttributeSet {
+	a.Attributes[attributeType] = NewAttribute(attributeType, value)
```
Because this function uses `NewAttribute`, we should ensure the function comments warns about the situations in which it panics.

jls5177 replied:
Makes sense, I've updated the few functions that call `NewAttribute`.

dmjones commented on attributes.go:
```diff
@@ -0,0 +1,256 @@
+}
+
+// keyPairAttributeHasID ensures both sets of attributes has a CkaId attribute
+func validateKeyPairAttributes(public, private *AttributeSet) error {
```
This function seems to require keys to have "public" or "private" in their CKA_ID attribute. This is not something we should enforce; perhaps this is code accidentally added from your own use case?

jls5177 replied:
Sorry, I am still learning the PKCS#11 spec. The previous code made it seem like `CkaId` and/or `CkaLabel` was required when making these API calls. I removed the additional checks from the various `Generate...WithAttributes()` functions.

dmjones commented on dsa.go:
```diff
@@ -79,11 +79,11 @@ func (c *Context) GenerateDSAKeyPair(id []byte, params *dsa.Parameters) (Signer,
 	}
-
-	return c.generateDSAKeyPair(id, nil, params)
+	return c.GenerateDSAKeyPairWithAttributes(template, template.Copy(), params)
```
Assuming we get rid of `validateKeyPairAttributes`, there should be no need to copy the template here. (This comment applies to all similar instances in the generation functions).

jls5177 replied:
The copy is required as the `Public` attribute set has different attributes from the `Private`. If we do not peform the copy then the API will fail. I will look into refactoring the code to make this clearer.

dmjones commented on dsa.go:
```diff
@@ -93,55 +93,47 @@ func (c *Context) GenerateDSAKeyPairWithLabel(id, label []byte, params *dsa.Para
+// private will be merged into the default templates used when generating keys. Where conflicts occur, the user
+// supplied attributes will win.
+func (c *Context) GenerateDSAKeyPairWithAttributes(public, private *AttributeSet, params *dsa.Parameters) (k *pkcs11PrivateKeyDSA, err error) {
+	if err = validateKeyPairAttributes(public, private); err != nil {
```
This function (and the other `...WithAttributes` functions) need this at the start:

```go
	if c.closed.Get() {
		return nil, errClosed
	}
```

jls5177 replied:
Good catch. I will add this to the various functions.

dmjones commented on dsa.go:
```diff
@@ -93,55 +93,47 @@ func (c *Context) GenerateDSAKeyPairWithLabel(id, label []byte, params *dsa.Para
 		g := params.G.Bytes()
 
-		publicKeyTemplate := []*pkcs11.Attribute{
+		public.AddDefaultAttributes([]*pkcs11.Attribute{
```
It might be polite to take a copy of these templates before we mutate them. Or... we see this as a feature and explain in the function comments that the final applied template is "returned" in the calling arguments. What do you think?

(This comment applies to all the `...WithAttributes` functions).

dmjones replied:
Sorry, I've just remembered your introduction:

>Removed the code that returned the modified Attribute slices since the new AttributeSet is mutable and will contain the added Attributes after the call.

So that implies we will comment this as a feature, which I'm happy with :-)

jls5177 replied:
The original commit was trying to return the modified AttributeSets. I felt this was very hard to maintain so I decided it would be better to just update the received AttributeSet so the caller can manipulate for subsequent calls if they needed to. 

I can make a clone instead and just not return the modified `AttributeSet`. What method would you prefer?

dmjones commented on dsa.go:
```diff
@@ -93,55 +93,47 @@ func (c *Context) GenerateDSAKeyPairWithLabel(id, label []byte, params *dsa.Para
+// GenerateDSAKeyPairWithAttributes creates a DSA key pair on the token. Additional attributes from public and
+// private will be merged into the default templates used when generating keys. Where conflicts occur, the user
+// supplied attributes will win.
+func (c *Context) GenerateDSAKeyPairWithAttributes(public, private *AttributeSet, params *dsa.Parameters) (k *pkcs11PrivateKeyDSA, err error) {
```
The return values should be _unnamed_ and using the public interface type, e.g. `(Signer, error)`. This same comment applies to the other `...WithAttributes` functions.

jls5177 replied:
I just noticed this same issue when I updated my codebase to use a version of this PR. Yes, I will change these to return the `Signer` interface instead.

jls5177 replied:
I switched the return type to `Signer` but left it named as the bulk of the logic is within the anonymous function passed to `withSession()` and the named argument allows the value to be easily returned from the outer function.

dmjones commented on symmetric.go:
```diff
@@ -273,26 +273,29 @@ func (c *Context) GenerateSecretKeyWithLabel(id, label []byte, bits int, cipher
+// private will be merged into the default templates used when generating keys. Where conflicts occur, the user
+// supplied attributes will win.
+func (c *Context) GenerateSecretKeyWithAttributes(template *AttributeSet, bits int, cipher *SymmetricCipher) (k *SecretKey, err error) {
+	if err = attributeHasID(template, "secret"); err != nil {
```
Per the comments above (in `attributes.go`), I think this check ought to be removed.

jls5177 replied:
Done

dmjones commented on attributes.go:
```diff
@@ -0,0 +1,256 @@
+	"github.com/miekg/pkcs11"
+)
+
+type AttributeType = uint
```
With the exception of the `CkaFoo` constants, exported types and functions must have comments. (I thought the CI checked this, but obviously not). This comment applies throughout this file (only mentioned once).

jls5177 replied:
I've added comments to the various types and functions.

dmjones commented on attributes.go:
```diff
@@ -0,0 +1,256 @@
+func (a *AttributeSet) ToSlice() []*Attribute {
+	var attributes []*Attribute
+	for _, v := range a.Attributes {
+		attributes = append(attributes, v)
```
Is a shallow copy OK here? If so, let's make that clear when adding a function comment.

jls5177 replied:
I converted this to a deep copy to match the `Copy()` function.

dmjones reviewed (changes requested):
Thank you for taking the time to finish this work. I've added quite a few review comments, but they are mostly tidy up work.

The two actual problems (which I've also added comments for) are:
- I would like to address the module-protected login issue separately
- Validation that keys have IDs with "public" and "private" is incorrect in the general use case

dmjones commented on attributes.go:
```diff
@@ -0,0 +1,256 @@
+type Attribute = pkcs11.Attribute
+
+// An AttributeSet groups together operations that are common for a slice of Attributes
+type AttributeSet struct {
```
We could just make this a map, for simplicity.

E.g.

```go
type AttributeSet map[AttributeType]*Attribute

func NewAttributeSet() AttributeSet {
	return make(AttributeSet)
}
```

The receiver functions would need to drop from `*AttributeSet` to `AttributeSet`. What do you think?

jls5177 replied:
Yeah this makes sense and actually makes the garbage collection faster since it does not have to track all of the pointers being passed around.

jls5177 committed:
refactor: Attribute comments and function names

Also, remove the explicit requirement for the various
Generate..WithAttribute() functions to have certain Attributes. This
will be up to the caller to verify they have the appropriate
attributes for the operation.

jls5177 committed:
feat: recover from panic from call to NewAttribute()

This change adds a recovery mechanism to the pkcs11.NewAttribute()
function to prevent from bringing down the application. Instead, the
error is captured and returned in an error variable. This required
all of the callers to now handle the returned error.

In addition, this change cleaned up the way the templates were
being duplicated to create the public/private sets. It should be
obvious why we are cloning the AttributeSets.

jls5177 committed:
cleanup: Attribute copy code

Adding a CopyAttribute() function that create a deep copy of an
Attribute without calling into NewAttribute which can panic. This
makes the Copy code cleaner as it does not have to worry about
handling errors.

jls5177 committed:
refactor: flatten AttributeSet to a map type

This cleans up the code and removes the need to pass around pointers.
